cmd/sink/main.go

```diff
@@ -50,46 +60,75 @@ func NewSink(db *sql.DB, logger *log.Logger) *Sink {
 	}
 
 	return &Sink{
+		ClusterName: clusterName,
 		Db:          db,
 		EventClient: eventClient,
 		Logger:      logger,
+		Lock:        &sync.Mutex{},
 	}
 }
 
 // Processes incoming cloudevents and writes them to the database
-func (sink Sink) Receive(event cloudevents.Event) {
+func (sink *Sink) Receive(event cloudevents.Event) {
 	sink.Logger.Println("received CloudEvent: ", event.ID())
-	archiveEntry, err := NewArchiveEntry(event)
+	archiveEntry, err := models.ResourceEntryFromCloudevent(event, sink.ClusterName)
 	if err != nil {
 		sink.Logger.Printf("cloudevent %s is malformed and will not be processed: %s\n", event.ID(), err)
 		return
 	}
 	sink.Logger.Printf("cloudevent %s contains all required fields. Attempting to write it to the database\n", event.ID())
+	sink.Lock.Lock()
+	defer sink.Lock.Unlock()
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	sink.Logger.Printf("checking if resource from cloudevent %s has been written to the database already\n", event.ID())
+	entryId, err := sink.Db.QueryResourceId(ctx, archiveEntry)
```

Comment:
You are querying the database to see if the resource already exists before writing it?  This is very inefficient, resulting in twice as many database queries as needed.

You should look at using an `upsert` instead.

See https://www.postgresql.org/docs/current/sql-insert.html.

Reply:
Also see https://github.com/kubearchive/kubearchive/issues/172, which may be required for the `upsert`.

Reply:
done!